test/pytest/test_browser.py

```diff
@@ -18,50 +20,97 @@
+    if 'COVERAGE_RCFILE' in os.environ:
+        coverage = ['coverage', 'run', '--parallel-mode', '--module']
+    else:
+        coverage = []
+
+    # pass the address through a separate fd, so that we can see g_debug() messages (which go to stdout)
+    addr_r, addr_w = os.pipe()
+    try:
+        server = subprocess.Popen(
+            [f'{BUILDDIR}/test-server', 'python3', '-m', *coverage, 'cockpit.bridge'],
+            env={**os.environ, 'TEST_SERVER_ADDRESS_FD': f'{addr_w}'},
+            stdin=subprocess.DEVNULL, pass_fds=(addr_w,), close_fds=True
+        )
+    except FileNotFoundError:
+        pytest.skip('No test-server')
+    os.close(addr_w)
+    address = os.read(addr_r, 1000).decode()
+    os.close(addr_r)
 
-    for result in glob.glob(f'{prefix}{fnmatch}', recursive=recursive):
-        assert result.startswith(prefix)
-        yield result[prefixlen:]
+    yield URL(address)
 
+    server.kill()
+    server.wait()
 
-@pytest.mark.parametrize('html', glob_py310('**/test-*.html', root_dir=f'{SRCDIR}/qunit', recursive=True))
-def test_browser(html):
-    if not os.path.exists(f'{BUILDDIR}/test-server'):
-        pytest.skip('no test-server')
+
+@contextlib.asynccontextmanager
+async def webdriver() -> AsyncIterator[WebdriverBidi]:
+    driver = ChromiumBidi(headless=os.environ.get('TEST_SHOW_BROWSER', '0') == '0')
+    await driver.start_session()
+    try:
+        yield driver
+    finally:
+        await driver.close()
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize('html', glob.glob('**/test-*.html', root_dir=f'{SRCDIR}/qunit', recursive=True))
+async def test_browser(html: str) -> None:
     if html in SKIP:
         pytest.skip()
     elif html in XFAIL:
         pytest.xfail()
 
-    if 'COVERAGE_RCFILE' in os.environ:
-        coverage = ['coverage', 'run', '--parallel-mode', '--module']
-    else:
-        coverage = []
+    async with spawn_test_server() as base_url, webdriver() as browser:
+        await browser.bidi(
+            'browsingContext.navigate',
+            context=browser.context,
+            url=str(base_url / 'qunit' / html),
+            wait='complete'
+        )
+
+        ignore_resource_errors = False
+        error_message = None
+
+        async for message in browser.logs:
+            if message.type == 'console':
+                if message.text == 'cockpittest-tap-done':
+                    break
+                elif message.text == 'cockpittest-tap-error':
+                    error_message = message.text
+                    break
+                elif message.text == 'cockpittest-tap-expect-resource-error':
+                    ignore_resource_errors = True
+                    continue
+                elif message.text.startswith('not ok'):
+                    error_message = message.text
+
+            elif message.type == 'warning':
+                print('WARNING', message.text)
+
+            else:
+                print('OTHER', message.type, message.args, message.text)
```

Comment:
Are these debugging leftovers? (1) we didn't do that in tap-cdp, and (2) webdriver_bidi *already* pints console messages to stderr, so isn't that duplicated then?

Reply:
meh?  I sort of like using print() in pytest because of the stdio redirection.  You'll only see this in case the test fails, and in that case, it'll probably be useful information.

About the repetition: I like that this gets shown "more loudly", since it could well be the thing causing the test to fail.  That wouldn't be at all obvious from the "normal" output.

Reply:
Not a biggie, I was just wondering how these are different/better than the already existing print()s to stderr.